### DISCUSSION TITLE
NX monorepo: Deploy and Run Node microservices locally with Docker and Kubernetes

### DISCUSSION THREAD

**Discussion by creotip** [nrwl/nx#8432] [Show and Tell]
Step by step guide I wrote on NX, Docker and Kubernetes

https://creotip.io/posts/nx-monorepo-running-microservices-locally-with-docker-kubernetes

**Comment by fider** [comment]
1. What about dependencies? I don't want to have extra 920MB of dependencies installed in image A, B, C (...and 90 others) only because one of developers required some dependencies for service X. **Any suggestion how to bypass this? How to install dependencies required only by service A when deploying service A?**
2. 
  ```
  # dependencies that express needs
  RUN npm install reflect-metadata tslib rxjs express
  ```
  Imo. not production ready solution. **Any suggestions on how to avoid this?** `implicitDependencies` in `project.json` or `dependencies` in `package.json` approach?

**Reply by juristr** [reply]
Since Nx has an internal project graph (which also happens to include npm packages), it perfectly knows which packages are used by a given project. As such you can generate a package.json during the build which can be used for installing only those packages required by the given project in the monorepo. 

Many of the built-in Nx executors have this already built-in. Look for the `generatePackageJson` flag or similar.

Also, there's a potentially helpful discussion here how to write your own executor to do exactly that:
https://github.com/nrwl/nx-labs/issues/31

**Reply by BBlackwo** [reply]
+1 for for the 2nd issue you mentioned @fider. `generatePackageJson` doesn't pick up some dependencies such as `reflect-metadata` (or `mysql` when using `typeorm`). A different workaround is to explicitly import those packages in your `main.ts` and then they will be added to the generated `package.json`:

```ts
import 'mysql';
import 'reflect-metadata';
```

BTW the official [Nx and Node Microservices](https://blog.nrwl.io/nx-and-node-microservices-b6df3cd1bad6) guide uses an explicit install in the Dockerfile, same as @creotip:

```Dockerfile
# dependencies that nestjs needs
RUN npm install reflect-metadata tslib rxjs @nestjs/platform-express
```

There's also this related issue #1777 (package.json per app)
